File: src/llm_cal/benchmark/runner.py
"""Benchmark runner — validate llm-infer-cal's output against curated references.

For each entry in dataset.yaml, run the evaluator against the model, then
compare each `expectations[]` field with the predicted value. Report a
table of pass/fail per check, plus a summary.

This is NOT a synthetic benchmark. Every expected value cites a source
(HF API, model card text, vLLM recipe, hand computation) so users can
audit.
"""

from __future__ import annotations

from dataclasses import dataclass
from functools import lru_cache
from importlib.resources import files
from pathlib import Path
from typing import Literal

from pydantic import BaseModel, Field

from llm_cal.common.yaml_loader import load_yaml
from llm_cal.core.evaluator import EvaluationReport, Evaluator

Status = Literal["PASS", "FAIL", "SKIP"]


class Expectation(BaseModel):
    field: str
    # Exactly one of these is used depending on `field`
    expected: str | int | bool | None = None
    expected_min: int | None = None
    expected_max: int | None = None
    source: str


class BenchmarkEntry(BaseModel):
    name: str
    model_id: str
    gpu: str
    engine: str = "vllm"
    expectations: list[Expectation] = Field(default_factory=list)


class BenchmarkDataset(BaseModel):
    schema_version: int
    entries: list[BenchmarkEntry]


@dataclass(frozen=True)
class CheckResult:
    entry_name: str
    field: str
    status: Status
    predicted: str
    expected: str
    source: str
    note: str | None = None


def _default_dataset_path() -> Path:
    return Path(str(files("llm_cal.benchmark").joinpath("dataset.yaml")))


@lru_cache(maxsize=1)
def load_dataset(path: Path | None = None) -> BenchmarkDataset:
    return load_yaml(path or _default_dataset_path(), BenchmarkDataset)


def run_all(
    evaluator: Evaluator | None = None,
    dataset: BenchmarkDataset | None = None,
) -> list[CheckResult]:
    """Run every check in the dataset. Returns flat list of results."""
    evaluator = evaluator or Evaluator()
    dataset = dataset or load_dataset()
    results: list[CheckResult] = []
    for entry in dataset.entries:
        try:
            report = evaluator.evaluate(
                model_id=entry.model_id,
                gpu=entry.gpu,
                engine=entry.engine,
            )
        except Exception as e:
            for exp in entry.expectations:
                results.append(
                    CheckResult(
                        entry_name=entry.name,
                        field=exp.field,
                        status="SKIP",
                        predicted="(evaluation failed)",
                        expected=_fmt_expected(exp),
                        source=exp.source,
                        note=f"{type(e).__name__}: {e}",
                    )
                )
            continue
        for exp in entry.expectations:
            results.append(_check_one(entry.name, report, exp))
    return results


def _check_one(entry_name: str, report: EvaluationReport, exp: Expectation) -> CheckResult:
    predicted_str, status = _evaluate_field(report, exp)
    return CheckResult(
        entry_name=entry_name,
        field=exp.field,
        status=status,
        predicted=predicted_str,
        expected=_fmt_expected(exp),
        source=exp.source,
    )


def _evaluate_field(report: EvaluationReport, exp: Expectation) -> tuple[str, Status]:
    """Return (predicted_str, PASS/FAIL/SKIP) for this field.

    Each `field` name matches a documented check in dataset.yaml.
    """
    if exp.field == "attention_variant":
        attn_actual = report.profile.attention.variant if report.profile.attention else "(none)"
        return attn_actual, ("PASS" if attn_actual == exp.expected else "FAIL")

    if exp.field == "quantization":
        quant_actual = report.weight.quantization_guess.value
        return quant_actual, ("PASS" if quant_actual == exp.expected else "FAIL")

    if exp.field == "is_moe":
        actual_bool = report.profile.is_moe
        return str(actual_bool), ("PASS" if actual_bool == exp.expected else "FAIL")

    if exp.field == "weight_bytes":
        actual_int = report.weight.total_bytes.value
        low = exp.expected_min or 0
        high = exp.expected_max or (1 << 62)
        passed = low <= actual_int <= high
        return f"{actual_int:,}", ("PASS" if passed else "FAIL")

    if exp.field == "fleet_prod_gpus":
        if report.fleet is None:
            return "(no fleet)", "SKIP"
        prod = next((o for o in report.fleet.options if o.tier == "prod"), None)
        if prod is None:
            return "(no prod tier)", "SKIP"
        passed = prod.gpu_count == exp.expected
        return str(prod.gpu_count), ("PASS" if passed else "FAIL")

    if exp.field == "fleet_prod_gpus_at_most":
        if report.fleet is None:
            return "(no fleet)", "SKIP"
        prod = next((o for o in report.fleet.options if o.tier == "prod"), None)
        if prod is None:
            return "(no prod tier)", "SKIP"
        passed = prod.gpu_count <= int(exp.expected or 0)
        return f"{prod.gpu_count} (max {exp.expected})", ("PASS" if passed else "FAIL")

    return "(unknown field)", "SKIP"


def _fmt_expected(exp: Expectation) -> str:
    if exp.expected is not None:
        return str(exp.expected)
    if exp.expected_min is not None or exp.expected_max is not None:
        lo = f"{exp.expected_min:,}" if exp.expected_min is not None else "-∞"
        hi = f"{exp.expected_max:,}" if exp.expected_max is not None else "+∞"
        return f"[{lo}, {hi}]"
    return "(unspecified)"


def render_results_text(results: list[CheckResult]) -> str:
    out = [
        "Benchmark results",
        "entry | field | predicted | expected | status",
    ]
    current_entry = None
    for r in results:
        entry_cell = r.entry_name if r.entry_name != current_entry else ""
        current_entry = r.entry_name
        out.append(
            f"{entry_cell} | {r.field} | {r.predicted} | {r.expected} | {r.status}"
        )

    total = len(results)
    passed = sum(1 for r in results if r.status == "PASS")
    failed = sum(1 for r in results if r.status == "FAIL")
    skipped = sum(1 for r in results if r.status == "SKIP")
    out.append(f"Total: {total}   PASS: {passed}   FAIL: {failed}   SKIP: {skipped}")

    if failed > 0:
        out.append(
            "Failures show the tool's prediction diverges from a curated source. "
            "Check the source column for the expected-value provenance."
        )

    return "\n".join(out)


def exit_code_from(results: list[CheckResult]) -> int:
    """0 if all PASS or only SKIP; 1 if any FAIL."""
    return 1 if any(r.status == "FAIL" for r in results) else 0

File: crates/llm-infer-cal-core/src/benchmark/runner.rs
use serde::Deserialize;

use crate::core::evaluator::{EvaluationReport, Evaluator};

const DATASET_YAML: &str = include_str!("../../../../src/llm_cal/benchmark/dataset.yaml");

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Status {
    Pass,
    Fail,
    Skip,
}

impl Status {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Pass => "PASS",
            Self::Fail => "FAIL",
            Self::Skip => "SKIP",
        }
    }
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
#[serde(untagged)]
pub enum ExpectedValue {
    String(String),
    Int(i64),
    Bool(bool),
}

impl std::fmt::Display for ExpectedValue {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ExpectedValue::String(value) => f.write_str(value),
            ExpectedValue::Int(value) => write!(f, "{value}"),
            ExpectedValue::Bool(true) => f.write_str("True"),
            ExpectedValue::Bool(false) => f.write_str("False"),
        }
    }
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
pub struct Expectation {
    pub field: String,
    #[serde(default)]
    pub expected: Option<ExpectedValue>,
    #[serde(default, deserialize_with = "deserialize_optional_u64")]
    pub expected_min: Option<u64>,
    #[serde(default, deserialize_with = "deserialize_optional_u64")]
    pub expected_max: Option<u64>,
    pub source: String,
}

impl Expectation {
    pub fn expected(field: &str, expected: ExpectedValue, source: &str) -> Self {
        Self {
            field: field.to_string(),
            expected: Some(expected),
            expected_min: None,
            expected_max: None,
            source: source.to_string(),
        }
    }

    pub fn range(
        field: &str,
        expected_min: Option<u64>,
        expected_max: Option<u64>,
        source: &str,
    ) -> Self {
        Self {
            field: field.to_string(),
            expected: None,
            expected_min,
            expected_max,
            source: source.to_string(),
        }
    }
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
pub struct BenchmarkEntry {
    pub name: String,
    pub model_id: String,
    pub gpu: String,
    #[serde(default = "default_engine")]
    pub engine: String,
    #[serde(default)]
    pub expectations: Vec<Expectation>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
pub struct BenchmarkDataset {
    pub schema_version: u64,
    pub entries: Vec<BenchmarkEntry>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct CheckResult {
    pub entry_name: String,
    pub field: String,
    pub status: Status,
    pub predicted: String,
    pub expected: String,
    pub source: String,
    pub note: Option<String>,
}

impl CheckResult {
    pub fn new(
        entry_name: &str,
        field: &str,
        status: Status,
        predicted: &str,
        expected: &str,
        source: &str,
    ) -> Self {
        Self {
            entry_name: entry_name.to_string(),
            field: field.to_string(),
            status,
            predicted: predicted.to_string(),
            expected: expected.to_string(),
            source: source.to_string(),
            note: None,
        }
    }

    fn with_note(mut self, note: String) -> Self {
        self.note = Some(note);
        self
    }
}

pub fn load_dataset() -> Result<BenchmarkDataset, serde_yaml::Error> {
    serde_yaml::from_str(DATASET_YAML)
}

pub fn run_all(evaluator: &Evaluator, dataset: &BenchmarkDataset) -> Vec<CheckResult> {
    let mut results = Vec::new();

    for entry in &dataset.entries {
        let report = evaluator.evaluate(
            &entry.model_id,
            &entry.gpu,
            &entry.engine,
            Default::default(),
        );
        match report {
            Ok(report) => {
                results.extend(
                    entry
                        .expectations
                        .iter()
                        .map(|expectation| check_one(&entry.name, &report, expectation)),
                );
            }
            Err(error) => {
                let note = format!("{error}");
                results.extend(entry.expectations.iter().map(|expectation| {
                    CheckResult::new(
                        &entry.name,
                        &expectation.field,
                        Status::Skip,
                        "(evaluation failed)",
                        &fmt_expected(expectation),
                        &expectation.source,
                    )
                    .with_note(note.clone())
                }));
            }
        }
    }

    results
}

pub fn check_one(
    entry_name: &str,
    report: &EvaluationReport,
    expectation: &Expectation,
) -> CheckResult {
    let (predicted, status) = evaluate_field(report, expectation);
    CheckResult::new(
        entry_name,
        &expectation.field,
        status,
        &predicted,
        &fmt_expected(expectation),
        &expectation.source,
    )
}

pub fn evaluate_field(report: &EvaluationReport, expectation: &Expectation) -> (String, Status) {
    match expectation.field.as_str() {
        "attention_variant" => {
            let actual = report
                .profile
                .attention
                .as_ref()
                .map(|attention| attention.variant.as_str())
                .unwrap_or("(none)");
            (
                actual.to_string(),
                pass_if_string(actual, expectation.expected.as_ref()),
            )
        }
        "quantization" => {
            let actual = report.weight.quantization_guess.value.as_str();
            (
                actual.to_string(),
                pass_if_string(actual, expectation.expected.as_ref()),
            )
        }
        "is_moe" => {
            let actual = report.profile.is_moe();
            (
                actual.to_string(),
                pass_if_bool(actual, expectation.expected.as_ref()),
            )
        }
        "weight_bytes" => {
            let actual = report.weight.total_bytes.value;
            let low = expectation.expected_min.unwrap_or(0);
            let high = expectation.expected_max.unwrap_or(1_u64 << 62);
            (
                fmt_u64(actual),
                if (low..=high).contains(&actual) {
                    Status::Pass
                } else {
                    Status::Fail
                },
            )
        }
        "fleet_prod_gpus" => {
            let Some(prod) = prod_option(report) else {
                return fleet_missing(report);
            };
            let passed = expectation
                .expected
                .as_ref()
                .and_then(expected_as_i64)
                .is_some_and(|expected| prod.gpu_count as i64 == expected);
            (
                prod.gpu_count.to_string(),
                if passed { Status::Pass } else { Status::Fail },
            )
        }
        "fleet_prod_gpus_at_most" => {
            let Some(prod) = prod_option(report) else {
                return fleet_missing(report);
            };
            let expected = expectation
                .expected
                .as_ref()
                .and_then(expected_as_i64)
                .unwrap_or(0);
            (
                format!("{} (max {expected})", prod.gpu_count),
                if prod.gpu_count as i64 <= expected {
                    Status::Pass
                } else {
                    Status::Fail
                },
            )
        }
        _ => ("(unknown field)".to_string(), Status::Skip),
    }
}

pub fn fmt_expected(expectation: &Expectation) -> String {
    if let Some(expected) = &expectation.expected {
        return expected.to_string();
    }

    if expectation.expected_min.is_some() || expectation.expected_max.is_some() {
        let low = expectation
            .expected_min
            .map(fmt_u64)
            .unwrap_or_else(|| "-∞".to_string());
        let high = expectation
            .expected_max
            .map(fmt_u64)
            .unwrap_or_else(|| "+∞".to_string());
        return format!("[{low}, {high}]");
    }

    "(unspecified)".to_string()
}

pub fn exit_code_from(results: &[CheckResult]) -> i32 {
    if results.iter().any(|result| result.status == Status::Fail) {
        1
    } else {
        0
    }
}

fn pass_if_string(actual: &str, expected: Option<&ExpectedValue>) -> Status {
    if matches!(expected, Some(ExpectedValue::String(expected)) if expected == actual) {
        Status::Pass
    } else {
        Status::Fail
    }
}

fn pass_if_bool(actual: bool, expected: Option<&ExpectedValue>) -> Status {
    if matches!(expected, Some(ExpectedValue::Bool(expected)) if *expected == actual) {
        Status::Pass
    } else {
        Status::Fail
    }
}

fn expected_as_i64(expected: &ExpectedValue) -> Option<i64> {
    match expected {
        ExpectedValue::Int(value) => Some(*value),
        ExpectedValue::String(value) => value.parse().ok(),
        ExpectedValue::Bool(_) => None,
    }
}

fn prod_option(report: &EvaluationReport) -> Option<&crate::fleet::planner::FleetOption> {
    report
        .fleet
        .as_ref()?
        .options
        .iter()
        .find(|option| option.tier == "prod")
}

fn fleet_missing(report: &EvaluationReport) -> (String, Status) {
    if report.fleet.is_none() {
        ("(no fleet)".to_string(), Status::Skip)
    } else {
        ("(no prod tier)".to_string(), Status::Skip)
    }
}

fn fmt_u64(value: u64) -> String {
    let text = value.to_string();
    let mut out = String::with_capacity(text.len() + text.len() / 3);
    let first_group = text.len() % 3;
    for (idx, ch) in text.chars().enumerate() {
        if idx > 0 && (idx == first_group || (idx > first_group && (idx - first_group) % 3 == 0)) {
            out.push(',');
        }
        out.push(ch);
    }
    out
}

fn default_engine() -> String {
    "vllm".to_string()
}

fn deserialize_optional_u64<'de, D>(deserializer: D) -> Result<Option<u64>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    let value = Option::<serde_yaml::Value>::deserialize(deserializer)?;
    let Some(value) = value else {
        return Ok(None);
    };
    match value {
        serde_yaml::Value::Number(number) => number
            .as_u64()
            .ok_or_else(|| {
                serde::de::Error::custom(format!("expected non-negative integer, got {number}"))
            })
            .map(Some),
        serde_yaml::Value::String(text) => text
            .replace('_', "")
            .parse::<u64>()
            .map(Some)
            .map_err(serde::de::Error::custom),
        other => Err(serde::de::Error::custom(format!(
            "expected integer or numeric string, got {other:?}"
        ))),
    }
}
